File: navdialogs/src/main/java/ua/leonidius/navdialogs/SaveDialog.java
package ua.leonidius.navdialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.SimpleAdapter;
import android.widget.Spinner;
import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;

import java.io.File;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class SaveDialog extends NavigationDialog implements AlertDialog.OnClickListener {

    private EditText nameField;
    private Spinner encodingSpinner;
    private Model viewModel = null;
    private Initializer initializer = new Initializer();

    public static SaveDialog create(String defaultName, String defaultEncoding,  Callback callback) {
        return create(defaultName, defaultEncoding, Environment.getExternalStorageDirectory(), callback);
    }

    public static SaveDialog create(String defaultName, String defaultEncoding, File defaultDirectory, Callback callback) {
        SaveDialog dialog = new SaveDialog();
        dialog.initializer.defaultEncoding = defaultEncoding;
        dialog.initializer.defaultFileName = defaultName;
        dialog.initializer.defaultDir = defaultDirectory;
        dialog.initializer.callback = callback;
        return dialog;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (!getViewModel().initialized) {
            initializer.initialize();
            getViewModel().initialized = true;
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder adb = new AlertDialog.Builder(getActivity());
        adb.setTitle(R.string.save_as);
        adb.setPositiveButton(android.R.string.ok, this);

        LayoutInflater inflater = (LayoutInflater)getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View dialogView = inflater.inflate(R.layout.navdialogs_dialog_save_as, null, false);
        initView(dialogView, savedInstanceState);
        adb.setView(dialogView);

        openDir(getViewModel().currentDir);

        Dialog dialog = adb.create();
        setMaxHeight(dialog);
        return dialog;
    }

    @Override
    protected void initView(View dialogView, Bundle savedState) {
        super.initView(dialogView, savedState);

        nameField = dialogView.findViewById(R.id.saveasdialog_name);
        nameField.setText(getViewModel().fileName);

        encodingSpinner = dialogView.findViewById(R.id.saveasdialog_encoding_spinner);

        String[] availableEncodings = getViewModel().getAvailableEncodings();

        ArrayList<Map<String, Object>> data = new ArrayList<>();
        Map<String, Object> m;
        int currentEncodingPosition = 0;
        for (int i = 0; i < availableEncodings.length; i++) {
            String encoding = availableEncodings[i];
            m = new HashMap<>();
            m.put("name", encoding);
            if (encoding.equalsIgnoreCase(getViewModel().currentEncoding)) {
                currentEncodingPosition = i;
            }
            data.add(m);
        }
        String[] from = {"name"};
        int[] to = {android.R.id.text1};

        encodingSpinner.setAdapter(new SimpleAdapter(getActivity(), data, android.R.layout.simple_list_item_1, from, to));
        encodingSpinner.setSelection(currentEncodingPosition);
    }

    @Override
    public void onClick(DialogInterface dialog, int id) {
        File file = new File(getViewModel().currentDir.getPath() + File.separatorChar + nameField.getText());
        String encoding = getViewModel().getAvailableEncodings()[encodingSpinner.getSelectedItemPosition()];

        if (!file.exists()) {
            getViewModel().callback.call(file, encoding);
            return;
        }

        RewriteDialog.create(rewrite -> {
            if (rewrite) getViewModel().callback.call(file, encoding);
            else show(getFragmentManager(), "saveDialog");
        }).show(getFragmentManager(), "rewriteDialog");
    }

    @Override
    public Model getViewModel() {
        if (viewModel == null) {
            viewModel = new ViewModelProvider(this).get(Model.class);
        }
        return viewModel;
    }

    public static class Model extends NavDialogViewModel {
        boolean initialized = false;
        Callback callback;
        String fileName;
        String currentEncoding;

        private String[] availableEncodings = null;

        String[] getAvailableEncodings() {
            if (availableEncodings == null) {
                availableEncodings = new String[Charset.availableCharsets().size()];
                Charset.availableCharsets().keySet().toArray(availableEncodings);
            }
            return availableEncodings;
        }
    }

    private class Initializer {
        File defaultDir;
        String defaultFileName;
        String defaultEncoding;
        Callback callback;

        void initialize() {
            getViewModel().currentDir = defaultDir;
            getViewModel().fileName = defaultFileName;
            getViewModel().currentEncoding = defaultEncoding;
            getViewModel().callback = callback;
        }
    }

    public interface Callback {
        void call(File file, String encoding);
    }

}

File: navdialogs/src/main/java/ua/leonidius/navdialogs/SaveDialog.kt
package ua.leonidius.navdialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.SimpleAdapter
import android.widget.Spinner
import androidx.lifecycle.ViewModelProvider
import java.io.File
import java.nio.charset.Charset
import java.util.*

class SaveDialog : NavigationDialog(), DialogInterface.OnClickListener {

    private lateinit var nameField: EditText
    private lateinit var encodingSpinner: Spinner
    private lateinit var viewModel: Model
    private var initializerFunction : (() -> Unit)? = null

    companion object {

        fun create(defaultName: String = ".txt", defaultEncoding: String = "UTF-8",
                   defaultDirectory: File = Environment.getExternalStorageDirectory(),
                   callback: (File, String) -> Unit): SaveDialog {
            val dialog = SaveDialog()
            dialog.initializerFunction = {
                with(dialog.getViewModel()) {
                    this.fileName = defaultName
                    this.currentEncoding = defaultEncoding
                    this.currentDir = defaultDirectory
                    this.callback = callback
                }
            }
            return dialog
        }

    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        initializerFunction?.invoke()
        initializerFunction = null
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val adb = AlertDialog.Builder(activity)
        adb.setTitle(R.string.save_as)
        adb.setPositiveButton(android.R.string.ok, this)

        val inflater =  LayoutInflater.from(context)
        val dialogView = inflater.inflate(R.layout.navdialogs_dialog_save_as, null, false)
        initView(dialogView)
        adb.setView(dialogView)

        openDir(getViewModel().currentDir)

        val dialog = adb.create()
        setMaxHeight(dialog)
        return dialog
    }

    override fun initView(dialogView: View) {
        super.initView(dialogView)

        nameField = dialogView.findViewById(R.id.nameField)
        nameField.setText(getViewModel().fileName)

        encodingSpinner = dialogView.findViewById(R.id.encodingSpinner)

        val availableEncodings = getViewModel().getAvailableEncodings()

        val data = ArrayList<Map<String, Any>>()
        var m: MutableMap<String, Any>
        var currentEncodingPosition = 0
        for (i in availableEncodings.indices) {
            val encoding = availableEncodings[i]
            m = HashMap()
            m["name"] = encoding
            if (encoding.equals(getViewModel().currentEncoding, ignoreCase = true)) {
                currentEncodingPosition = i
            }
            data.add(m)
        }
        val from = arrayOf("name")
        val to = intArrayOf(android.R.id.text1)

        encodingSpinner.adapter = SimpleAdapter(activity, data, android.R.layout.simple_list_item_1, from, to)
        encodingSpinner.setSelection(currentEncodingPosition)
    }

    override fun onClick(dialog: DialogInterface, id: Int) {
        val file = File(getViewModel().currentDir!!.path + File.separatorChar + nameField.text)
        val encoding = getViewModel().getAvailableEncodings()[encodingSpinner.selectedItemPosition]

        if (!file.exists()) {
            getViewModel().callback(file, encoding)
            return
        }

        RewriteDialog.create { rewrite ->
            if (rewrite)
                getViewModel().callback(file, encoding)
            else
                show(parentFragmentManager, "saveDialog")
        }.show(parentFragmentManager, "rewriteDialog")
    }

    override fun getViewModel(): Model {
        if (!this::viewModel.isInitialized) {
            viewModel = ViewModelProvider(this).get(Model::class.java)
        }
        return viewModel
    }

    class Model : NavigationDialog.NavDialogViewModel() {
        internal lateinit var callback: ((File, String) -> Unit)
        internal lateinit var fileName: String
        internal lateinit var currentEncoding: String

        private lateinit var availableEncodings: Array<String>

        internal fun getAvailableEncodings(): Array<String> {
            if (!this::availableEncodings.isInitialized) {
                availableEncodings = Charset.availableCharsets().keys.toTypedArray()
            }
            return availableEncodings
        }
    }

}